File: Exercise/Hard/04. Bracket Validator/Code/Bracket Validator.cs
using System;
using System.Collections.Generic;

class Program {
    static void Main() {
        string expression = Console.ReadLine();

        if (IsValidExpression(expression)) {
            Console.WriteLine("True");
        }
        else {
            Console.WriteLine("False");
        }
    }

    static bool IsValidExpression(string expression) {
        Stack<char> stack = new Stack<char>();

        foreach (char ch in expression) {
            if (IsOpeningBracket(ch)) {
                stack.Push(ch);
            }
            else {
                if (stack.Count == 0 || !IsValidPair(stack.Pop(), ch)) {
                    return false;
                }
            }
        }

        return stack.Count == 0;
    }

    static bool IsOpeningBracket(char ch) {
        return ch == '(' || ch == '{' || ch == '[';
    }

    static bool IsValidPair(char opening, char closing) {
        return (opening == '(' && closing == ')') ||
               (opening == '{' && closing == '}') ||
               (opening == '[' && closing == ']');
    }
}

File: Exercise/Hard/04. Bracket Validator/Code/Bracket Validator.py
def is_valid_expression(expression):
    stack = []
    opening_brackets = "({["
    closing_brackets = ")}]"

    for char in expression:
        if char in opening_brackets:
            stack.append(char)
        elif char in closing_brackets:
            if not stack or not is_valid_pair(stack.pop(), char):
                return False

    return len(stack) == 0

def is_valid_pair(opening, closing):
    return (opening == '(' and closing == ')') or \
           (opening == '{' and closing == '}') or \
           (opening == '[' and closing == ']')

def main():
    expression = input().strip()

    if is_valid_expression(expression):
        print("True")
    else:
        print("False")

if __name__ == "__main__":
    main()
